### PR TITLE
fix: race condition in directory app

### PR DESCRIPTION
**Pull request by konstantina-gkp**
fixes: #3324

When loading the initial page of the directory app, the filters would be available for use. If a filter was applied while the home page was still loading, the directory app would display all the biobanks and not the results matching the selected filter. Note: to test the initial issue you need to simulate a slower internet connection.
The issue is fixed by timestamping the requests. This is used to ensure that only the most recent request updates the page, preventing race conditions where older requests (e.g., loading the initial page) might overwrite newer ones (applying filters).
In addition, the functions getBiobankCards() and updateBiobankCards() have been refactored.

### CODE REVIEW

**Review by esthervanenckevort — APPROVED** [review]
Code change looks okay to me, but concurrency is always tricky. So needs one other frontend developer to have a look at it as well.

File: apps/directory/src/stores/biobanksStore.js [hunk 1]
```diff
@@ -102,34 +102,30 @@ export const useBiobanksStore = defineStore("biobanksStore", () => {
...
+      lastRequestTime = requestTime;
+
       const biobankResult = await baseQuery.execute();
 
-      /** depending on whether filters have been selected display the correct count of biobanks */
-      let foundBiobanks = biobankResult.Biobanks;
-      if (filtersStore.hasActiveFilters) {
-        foundBiobanks = foundBiobanks.filter((biobank) => biobank.collections);
+      /* Update biobankCards only if the result is the most recent one*/
+      if (requestTime === lastRequestTime) {
```
Comment (chinook25) [inline apps/directory/src/stores/biobanksStore.js:118]: Won't this always be true? I feel like line 113 should happen at the end on the function.

File: apps/directory/src/stores/biobanksStore.js [hunk 2]
```diff
@@ -102,34 +102,30 @@ export const useBiobanksStore = defineStore("biobanksStore", () => {
...
+        let foundBiobanks = biobankResult.Biobanks;
+        if (filtersStore.hasActiveFilters) {
+          foundBiobanks = foundBiobanks.filter(
```
Comment (chinook25) [inline apps/directory/src/stores/biobanksStore.js:121]: foundBiobanks is a too vague name, I think it should be biobanksWithCollections (if I read the code correctly).

File: apps/directory/src/stores/biobanksStore.js [hunk 3]
```diff
@@ -102,34 +102,30 @@ export const useBiobanksStore = defineStore("biobanksStore", () => {
     return facetBiobankColumnDetails;
   }
 
+  let lastRequestTime = 0;
```
Comment (chinook25) [inline apps/directory/src/stores/biobanksStore.js:105]: please define globals at the top of the file

File: apps/directory/src/stores/biobanksStore.js [hunk 4]
```diff
@@ -102,34 +102,30 @@ export const useBiobanksStore = defineStore("biobanksStore", () => {
...
+      const requestTime = Date.now();
+      lastRequestTime = requestTime;
+
       const biobankResult = await baseQuery.execute();
 
-      /** depending on whether filters have been selected display the correct count of biobanks */
-      let foundBiobanks = biobankResult.Biobanks;
-      if (filtersStore.hasActiveFilters) {
-        foundBiobanks = foundBiobanks.filter((biobank) => biobank.collections);
+      /* Update biobankCards only if the result is the most recent one*/
```
Comment (chinook25) [inline apps/directory/src/stores/biobanksStore.js:117]: comment is redundant, code is readable enough, so I don't think it needs further explanation ;)

**Review by chinook25 — CHANGES_REQUESTED** [review]

File: apps/directory/src/stores/biobanksStore.js (hunk 4 above)
Comment (konstantina-gkp) [reply apps/directory/src/stores/biobanksStore.js:117]: I think the comment should remain for clarity, since the code is slightly tricky.

**Review by chinook25 — APPROVED** [review]

**Pull request merged**